spec/schemas/security._common.yaml

```diff
@@ -292,69 +602,97 @@ components:
       additionalProperties:
-        $ref: '#/components/schemas/RoleMapping'
-    DynamicConfig:
+        $ref: '#/components/schemas/Tenant'
+
+    Unauthorized:
       type: object
       properties:
-        dynamic:
-          $ref: '#/components/schemas/DynamicOptions'
-    DynamicOptions:
+        status:
+          type: string
+          value: 403
+        message:
+          type: string
+          description: Message returned as part of FORBIDDEN response.
+
+    UpgradeCheck:
       type: object
       properties:
-        filteredAliasMode:
-          type: string
-        disableRestAuth:
-          type: boolean
-        disableIntertransportAuth:
-          type: boolean
-        respectRequestIndicesOptions:
-          type: boolean
-        kibana: {}
-        http: {}
-        authc: {}
-        authz: {}
-        authFailureListeners: {}
-        doNotFailOnForbidden:
-          type: boolean
-        multiRolespanEnabled:
-          type: boolean
-        hostsResolverMode:
+        status:
           type: string
-        doNotFailOnForbiddenEmpty:
+        upgradeAvailable:
           type: boolean
-    CertificatesDetail:
+        upgradeActions:
+          type: object
+
+    UpgradePerform:
       type: object
       properties:
-        issuer_dn:
+        status:
           type: string
-        subject_dn:
-          type: string
-        san:
-          type: string
-        not_before:
-          type: string
-        not_after:
-          type: string
-    TenantsMap:
-      type: object
-      additionalProperties:
-        $ref: '#/components/schemas/Tenant'
-    Tenant:
+        upgrades:
+          type: object
+
+    User:
       type: object
       properties:
+        hash:
+          type: string
         reserved:
           type: boolean
         hidden:
           type: boolean
+        backend_roles:
+          type: array
+          items:
+            type: string
+        attributes:
+          $ref: '#/components/schemas/UserAttributes'
         description:
           type: string
+        opendistro_security_roles:
+          type: array
+          items:
+            type: string
         static:
           type: boolean
-    CreateTenantParams:
+
+    UserAttributes:
+      type: object
+      additionalProperties:
+        type: string
+
+    UserTenants:
       type: object
       properties:
-        description:
+        global_tenant:
+          type: boolean
+        admin_tenant:
+          type: boolean
+        admin:
+          type: boolean
+
+    UsersMap:
+      type: object
+      additionalProperties:
+        $ref: '#/components/schemas/User'
+
+    WhoAmI:
+      type: object
+      properties:
+        dn:
+          type: string
+        is_admin:
+          type: string
+        is_node_certificate_request:
+          type: string
+
+    TenantInfo:
+      type: object
+      properties:
+        "<indexOrAlias>":
           type: string
+          value: "<tenant-name>"
```

Comment:
I think think you should use [additionalProperties](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties) here instead of `properties`. There are many of objects with this property in this repo you can use as examples.

Reply:
I know what you mean. In JSON schema however, `properties` describes fix/predefined keys (similar to a structure in Smithy) while `additionalProperties` describes free-form keys (similar to a map in Smithy). You can actually use both in an object to describe map with some reserved keys.

The schema
```yaml
TenantInfo:
      type: object
      properties:
        "<indexOrAlias>":
          type: string
          value: "<tenant-name>"
```
describes an object with exactly one key called `<indexOrAlias>` with value of type string (`value: "<tenant-name>"` will be ignored because `value` is not part of JSON schema)
You can either do:
```
TenantInfoMap:
      type: object
      additionalProperties:
            type: string        
```
which is basically a map with keys of type string (the default schema for keys) and value of type string as well.
Or if you really want to describe this entity with the highest resolution possible (though might not be necessary):
```yaml       
TenantInfoMap:
     additionalProperties:
          $ref: #/components/schemas/TenantName
     oneOf:
          - type: object
             propertyNames:
                  $ref: #/components/schemas/Index
          - type: object
             propertyNames:
                  $ref: #/components/schemas/Alias
TenantName:
     type: string
Index:
     type: string
Alias:
     type: string
```
This schema is identical to the one above as far as schema validation is concerned (basically a map with both keys and values as string) but it provides much more information about the meanings of the keys and the values. Plus, you can expand the `TenantName`, `Index` and `Alias` to give you info via `description` or `title` or add restrictions to them via `pattern`, `enum` and so on.

Reply:
> describes an object with exactly one key called <indexOrAlias> with value of type string 

in this case it could be more than 1 entries so, 
```
TenantInfo:
      type: object
      additionalProperties:
            type: string        
```
would suffice. Let me know if this needs more resolution.

Reply:
We can always up the resolution in the future or simply add a `description` for TenantInfo. We also have Generated Docs is our roadmap so any extra info added will help with that future project.